crates/noirc_evaluator/src/ssa/ssa_gen/context.rs

```diff
@@ -317,7 +317,7 @@ impl<'a> FunctionContext<'a> {
             rhs,
             &self.builder.current_function.dfg,
         ) {
-            let result_type = self.builder.current_function.dfg.type_of_value(result);
+            let result_type = self.builder.current_function.dfg.type_of_value(lhs);
```

Comment:
Shouldn't the type of the result value still be the type of the bit shift result we expect? Looks like we're still leaving it as a field here and just truncating the field afterward. Can we change `insert_shift_{left,right}` to possibly cast the result to the desired bit type instead?

Reply:
For me it is fine because all our binary operations have the same type as their lhs.
If we do not want to rely on this, we could return (result, result_type) from the first match and provide the correct type for shifts and the type of result for the others.

Reply:
We still use the type of the lhs in any simplification we do for binary ops as well

Reply:
> For me it is fine because all our binary operations have the same type as their lhs.

The issue is that this instance breaks that rule since the type of `result` here does not match the type of `lhs`, which can be confusing

Reply:
We're also multiplying and adding unsigned integers by fields in `pow` which isn't expected

Reply:
I have switched to casting the result inside of `insert_shift_{left,right}` as I see your point. This can provide a quick fix for the issue this PR set out to resolve, and then once we incorporate overflow protection we can just get rid of this truncation logic in general.